### ISSUE TITLE
Review representation of `mask` in `Event`

### ISSUE THREAD

**Issue by hannobraun** [hannobraun/inotify-rs#76]
From reading the man pages, I get the impression that distinct event objects are generated for each type of event, and that therefore most bits in the event mask can't coexist. If that is the case, maybe it makes sense to create a new enum, `EventKind`, to represent the type of event.

One exception to this case is `IN_ISDIR`, which can be set in addition to other bits that identify the type of event. A new `bool` field could be created specifically for this flag.

In any case, before anything like this can be merged, more careful reading of the man pages and probably some experimentation is required.

**Comment by hannobraun** [comment]
I just had this thought regarding this issue:
The current test suite covers the basics, assuming that anything not explicitely covered will work fine, since we don't do anything fance that's specific to the type of event. Before implementing this issue, we should have a full test suite in place that covers all types of events, to make sure all our assumptions are correct.

Extending the test suite in such a way would be a good idea anyway, but it would be especially important before we hardcode any assumptions about how the event mask is structured.

**Comment by hannobraun** [comment]
Another tought:
I think it would be better to model event queue overflows as an error, not just another kind of event.

**Comment by hannobraun** [comment]
Additional thoughts:
- The `cookie` could be removed. Instead it could be added as a field to the rename event variants of the `EventKind` enum. Then an `Eq` implementation or a dedicated method could be used to determine whether the rename events match.
- `IN_IGNORED` events could be made more explicit. Right now they're just another kind of event.

**Comment by lefuturiste** [comment]
I would like to have a representation of `EventType` as enums, it would simplify using match patterns.

**Comment by JackDyre** [comment]
As far as I understand, the `mask` returned from reading an event contains:

Exactly 0 or 1 of the following bitflags:
```
IN_ACCESS
IN_ATTRIB
IN_CLOSE_WRITE
IN_CLOSE_NOWRITE
IN_CREATE
IN_DELETE
IN_DELETE_SELF
IN_MODIFY
IN_MOVE_SELF
IN_MOVED_FROM
IN_MOVED_TO
IN_OPEN
```

0 or more of of the following bitflags:
```
IN_ISDIR
IN_IGNORED
IN_Q_OVERFLOW
IN_UNMOUNT
```

So it can be represented as an option of an enum of the primary bitflags, and a bitflag of the 4 auxillary bitflags.

**Comment by hannobraun** [comment]
Thank you for looking into this, @JackDyre!

What you say matches my understanding of inotify, but despite maintaining this library for more than 10 years, I'm definitely no expert 😄 

I think your suggestion of enum + bitflags sounds totally reasonable. I'd be happy to accept a pull request along those lines.

**Comment by hannobraun** [comment]
Whether or not we want to take additional measures on top of that (like the suggestions I made earlier in the thread) is a separate question. I'm always happy to merge anything that is a clear improvement, and we can always refine it later.

This issue has been open since 2017, so obviously there's no rush.

---

Although, regarding the immediate suggestion of enum + bigtflags, it's worth considering if we can do something to make the transition easier. Maybe have methods on `Event` that are similar to what `bitflags` provides.

**Comment by JackDyre** [comment]
> I think it would be better to model event queue overflows as an error, not just another kind of event.

I think this is a good idea. I have implemented this in my PR


>  The `cookie` could be removed. Instead it could be added as a field to the rename event variants of the `EventKind` enum. Then an `Eq` implementation or a dedicated method could be used to determine whether the rename events match.

I think this is a good idea, I will implement this tomorrow

>  `IN_IGNORED` events could be made more explicit. Right now they're just another kind of event.


As far as I understand, aren't `IN_IGNORED` events just normal events that let you know that the associated wd is no longer being monitored for future events?

**Comment by hannobraun** [comment]
> As far as I understand, aren't `IN_IGNORED` events just normal events that let you know that the associated wd is no longer being monitored for future events?

Yeah, that sounds about right. Not sure why I thought it could make sense to somehow make them more explicit. Let's just disregard that old comment.
